crates/circuit/src/dag_circuit.rs

```diff
@@ -4996,6 +4991,45 @@ impl DAGCircuit {
         Ok(new_node)
     }
 
+    fn _push_back(&mut self, py: Python, instr: PackedInstruction) -> PyResult<DAGNodeInfo> {
+        let op_name = instr.op.name();
+        let (all_clbits, vars): (Vec<Clbit>, Option<Vec<Var>>) = {
+            if self.may_have_additional_wires(py, &instr) {
+                let mut clbits: HashSet<Clbit> =
+                    HashSet::from_iter(self.cargs_interner.get(instr.clbits).iter().copied());
+                let (additional_clbits, additional_vars) = self
+                    .additional_wires(py, instr.op.view())
+                    .expect("Error trying to find additional wires for instruction");
+                for clbit in additional_clbits {
+                    clbits.insert(clbit);
+                }
+                let additional_vars = additional_vars
+                    .into_iter()
+                    .map(|var| self.vars.find(&var.bind(py).into()).unwrap())
+                    .collect();
+                (clbits.into_iter().collect(), Some(additional_vars))
+            } else {
+                (self.cargs_interner.get(instr.clbits).to_vec(), None)
+            }
+        };
```

Comment:
Instead of `_push_back`, I think it might be better to just pull this chunk out into its own method, `get_classical_resources(&PackedInstruction) -> (Vec<Clbit>, Option<Vec<Var>>)` and call it from `DAGCircuit::push_back` and `DAGCircuitBuilder::push_back`. The only other thing `_push_back` is doing (other than packing up where it left off) is incrementing the `op` count, which is fine to duplicate between the two `push_back`s, IMO.